### DISCUSSION TITLE
아이템 30. 이왕이면 제네릭 메서드로 만들라

### DISCUSSION THREAD

**Discussion by coalong** [Study-2-Effective-Java/Effective-Java#57] [3. 과제]
# 아이템 30. 이왕이면 제네릭 메서드로 만들어라

# ***Summary***

<aside>   

- 명시적으로 형변환해야 하는 메서드보다 제네릭 메서드가 더 안전하며 사용하기 쉽다.   

- 형변환을 해줘야 하는 기존 메서드를 제네릭하게 만들어 기존 클라이언트는 그대로 두고 새로운 사용자에게 편리함을 준다.

</aside>

# 0. 아이템 소개
제네릭 타입(아이템 29)에 연장하여 가능하면 메서드도 제네릭으로 만들어라 는 내용입니다.

제네릭 메서드를 사용했을 때 이점과 활용 예시를 보겠습니다.

# 1. 제네릭 메서드

- **`제네릭 메서드란?`** 
    - 매개 타입과 리턴 타입으로 타입 파라미터를 갖는 메서드 이다.
- **`제네릭이란?`** 데이터 타입을 일반화하는 것을 의미한다.
    - 클래스나 메서드를 정의할 때 데이터 타입을 변수로 지정하고, 사용할 때 타입을 정해줄 수 있는 것
- **`타입 파라미터란?`**
    - 변수로 지정한 데이터 타입이다.

```java
public <타입파라미터, ...> 리턴타입 메서드명(매개변수, ...) {...}
public <T> Box<T> boxing(T t) {...}
```

- 클래스와 마찬가지로(아이템 29), 메서드도 제네릭으로 만들 수 있다.
- 매개변수화 타입을 받는 정적 유틸리티 메서드는 보통 제네릭이다.

<details>
<summary>Collections의 `알고리즘` 메서드(binarySearch, sort 등) 는 모두 제네릭이다.</summary>
<div markdown="1">    

 [Java Docs](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/Collections.html)
        
```java
    // sort 
    @SuppressWarnings({"unchecked", "rawtypes"})
    public static <T> void sort(List<T> list, Comparator<? super T> c) {
        list.sort(c);
    }

    // binarySearch 
    public static <T> int binarySearch(List<? extends Comparable<? super T>> list, T key) {
        if (list instanceof RandomAccess || list.size()<BINARYSEARCH_THRESHOLD) 
            return Collections.indexedBinarySearch(list, key);
        else 
            return Collections.iteratorBinarySearch(list, key);
    }
```
</div>
</details>
   
        

## 1.1 문제가 있는 메서드

- **로 타입 사용 - 수용 불가!** (아이템 26)

```java
public static Set union(Set s1, Set s2) {
    Set result = new HashSet(s1);
    result.addAll(s2);
    return result;
}
```

```bash
[WARNING] uses unchecked or unsafe operations.
```

- 위와 같이 코드를 작성하면 **컴파일은 되지만 경고가 발생**한다.
- 경고를 없애려면 이 메서드를 타입 안전하게 만들어야 한다.
    
    **→** **제네릭 메서드로 수정!!!**
    

## 1.2 제네릭 메서드 작성법

1. 메서드 선언에서의 **세 집합(입력 2개, 반환 1개)의 원소 타입**을 **타입 파라미터로** 명시
2. 메서드 안에서도 이 **타입 파라미터만 사용하게 수정**하면 된다. 
**`(타입 파라미터들을 선언하는) 타입 파라미터 목록은 메서드의 접근 제어자와 반환 타입 사이에 온다.`**

```java
public static <E> Set<E> union(Set<E> s1, Set<E> s2) {
    Set<E> result = new HashSet<>(s1);
    result.addAll(s2);
    return result;
}
```
> 위의 코드는 (입력 2개, 반환 1개) 3개의 타입이 모두 같아야 한다.
한정적 와일드카드 타입(아이템 31)을 사용하여 유연성을 높일 수 있다.
> 
- 다음 코드에서 매개 변수 목록은 `<E>` 이고  반환 타입은 `Set<E>` 이다.   

<details>
<summary>타입 파라미터의 명명 규칙은 제네릭 타입과 같다.</summary>
<div markdown="1">        

- **타입 파라미터 명명 규칙**
    - 한문자로 이름 짓기
    - 대문자로 이름 짓기    
    
- **자주 사용하는 타입 인자**
    - E (Element)
    - K (Key)
    - N (Number)
    - T (Type)
    - V (Value)   
    
</div>
</details>

# 2. 제네릭 메서드 활용

- 제네릭 메서드를 활용한 쓰임과 `제네릭 싱글턴 팩터리`에 대해 살펴보자.

## 2.1 불변 객체를 여러 타입으로 활용하는 경우

- 제네릭은 **런타임에 타입 정보가 소거**(아이템28)되므로 하나의 객체를 **어떤 타입으로든 매개변수화**할 수 있다.
    - 제네릭의 타입 소거(erasure)란?
        - 컴파일 타임에만 타입에 대한 제약 조건을 검사하고, 런타임에는 해당 타입 정보를 알 수 없다.
          
- 그러나 요청한 타입 매개 변수에 맞게 **매번 그 객체의 타입을 바꿔주는 정적 팩터리**를 만들어야 한다. 이 패턴을 `제네릭 싱글턴 팩터리` 라 한다.    

<details>
<summary>제네릭 싱글턴 팩터리란?</summary>
<div markdown="1">           

- 제네릭으로 타입설정 가능한 인스턴스를 만들어주고, 반환 시에 제네릭으로 받은 타입을 이용해 타입을 결정하는 것이다.
    
```java
public class GenericFactoryMethod {
    public static final Set EMPTY_SET = new HashSet();
    
    public static  <T> Set<T> emptySet() {
        return (Set<T>) EMPTY_SET;
    }
    
    public static void main(String[] args) {
        Set<String> set1 = GenericFactoryMethod.emptySet();
        Set<Integer> set2 = GenericFactoryMethod.emptySet();
    
        set1.add("ab");
        set2.add(123);
    
        String s = set1.toString();
        System.out.println("s = " + s);
        // s = [ab, 123]
    }
}
```
> 유연하게 타입을 받을 수 있다.
 불변 객체를 타입에 맞게 여러번 재사용할 수 있다.
> 
- 제네릭 싱글턴 팩터리 예시
    - Collections.reverseOrder 같은 함수 객체 (아이템 42)
    - Collections.emptySet 같은 컬렉션
    
```java
// reverseOrder 
public static <T> Comparator<T> reverseOrder() {
// T 타입으로 캐스팅된 Singleton ReverseComparator 반환, 타입정보는 소거됨으로 그때그때 캐스팅만 되는 것
    return (Comparator<T>) ReverseComparator.REVERSE_ORDER; 
}
    
//emptySet
public static final Set EMPTY_SET = new EmptySet<>();
    
public static final <T> Set<T> emptySet() { 
    return (Set<T>) EMPTY_SET; 
}
```
</div>
</details>

## 2.2 항등함수(identity function)를 담은 클래스를 만드는 경우

- 항등 함수란?  **입력 값 수정없이 그대로 반환하는 함수**이다.
- 자바 라이브러리 Function.identity를 사용하면 되지만(아이템 59) 직접 한번 작성해보자.
- 항등함수 객체는 상태가 없으니 싱글톤으로 만드는 것이 좋다. 자바의 제네릭이 실체화된다면 항등항수를 타입별로 만들어야 하지만, 소거되므로 **제네릭 싱글턴**으로 충분하다.

```java
private static UnaryOperator<Object> IDENTITY_FN = (t) -> t;
/*
  항등함수란 입력값을 수정없이 그대로 반환하는 함수이므로
  T가 어떤 타입이든 UnaryOperator<T>를 사용해도 타입 안전하다.
  따라서, 비검사 형변환 경고를 숨겨서 사용해도 된다.
*/

@SuppressWarnings("unchecked") //비검사 형변환 경고 방지 (컴파일 시 경고 없이 완료된다.)
public static <T> UnaryOperator<T> identityFunction() {
	return (UnaryOperator<T>) IDENTITY_FN;
}
```

- 제네릭 싱글턴을 사용하여 UnaryOperator<String>과 UnaryOperator<Number>로 사용한 코드

```java
public static void main(String[] args) {

  // UnaryOperator<String>
  String[] strings = {"삼배", "대마", "나일론"};
  UnaryOperator<String> sameString = identityFunction();
  for (String s : strings) {
        System.out.println(sameString.apply(s));
}

  // UnaryOperator<Number>
  Number[] numbers = {1, 2.0, 3L};
  UnaryOperator<Number> sameNumber = identityFunction();
  for (Number n : numbers) {
        System.out.println(sameNumber.apply(n));
  }
```

## 2.3 재귀적 타입 한정(recursive type bound)의 사용

- 드문 경우이지만, **재귀적 타입 한정** 의 제네릭 사용
- 재귀적 타입 한정이란?
    - **자기 자신**이 들어간 **표현식**을 사용해 **타입 파라미터**의 **허용범위**를 **한정**
- 재귀적 타입 한정은 주로 타입의 자연적 순서를 정하는 Comparable 인터페이스와 함께 쓰인다.

```java
public interface Comparable<T> {
    int compareTo(T o);
}
// 타입 매개변수 T는 Comparable<T>를 구현한 타입이 비교할 수 있는 원소의 타입을 정의한다.
// 즉, T는 비교 가능한 타입이다. (자신과 같은 타입의 원소와만 비교할 수 있다.)
```

- Comparable을 구현한 컬레션의 **모든 원소가 상호 비교** 되어야 하는데, 
그 제약을 `<E extends Comparable <E>>` 라고 표현한다.

```java
public static <E extends Comparable<E>> E max(Collection<E> c) {
    if (c.isEmpty()) {
        // 빈 컬렉션을 건내면 예외를 던지니 Optional<E>를 반환하도록 수정하는 것이 나을 것(아이템 55)
	throw new IllegalArgumentException("Empty collection"); 
    }
    E result = null;
    for (E e : c){
	if (result == null || e.compareTo(result) > 0) {
	    result = Objects.requireNonNull(e);
	}
    return result;
    }
}
```

- `<E extends Comparable <E>>` 는 `모든 타입 E는 자신과 비교할 수 있다` 라는 뜻으로 해석한다.
    - 상호 비교 가능하다는 뜻을 아주 정확하게 표현한 것.

# *Reference*

[https://medium.com/depayse/kotlin-제네릭-generic-1-정의-타입-파라미터-제약-3bc8cf759cb3](https://medium.com/depayse/kotlin-%EC%A0%9C%EB%84%A4%EB%A6%AD-generic-1-%EC%A0%95%EC%9D%98-%ED%83%80%EC%9E%85-%ED%8C%8C%EB%9D%BC%EB%AF%B8%ED%84%B0-%EC%A0%9C%EC%95%BD-3bc8cf759cb3)

[https://blog.naver.com/PostView.naver?blogId=zzang9ha&logNo=222059024135&redirect=Dlog&widgetTypeCall=true&directAccess=false](https://blog.naver.com/PostView.naver?blogId=zzang9ha&logNo=222059024135&redirect=Dlog&widgetTypeCall=true&directAccess=false)

[https://jyami.tistory.com/99](https://jyami.tistory.com/99)

[https://velog.io/@tamxt4047/recirsove-type-bound](https://velog.io/@tamxt4047/recirsove-type-bound)

**Comment by JoisFe** [comment]
당연히 클래스, 인터페이스 뿐만 아니라 메서드 또한 로 타입으로 입력 매개변수, 리턴을 구현한다면
크게 두가지 문제
1. 특정한 타입만 사용할 수 있음 (일반성이 없음)
2. 타입 안정성이 없음 (컴파일 시점이 아닌 런타임 시점에서 예상치 못한 에러가..)

가 발생할 수 있으니 이번 아이템을 참고로 메서드를 구현해야하는 상황에서도 제네릭을 적용해 더 좋은 코드 작성을 위해 활용하도록 하겠습니다.
감사합니다.!!
